Create guides for registering LSP12 issued assets (#912)
lukso-network/docs: docs/learn/dapp-developer/register-lsp12-issued-assets.md

---
sidebar_label: 'Register LSP12 Issued Assets'
sidebar_position: 13

> description: Learn how to register Issued Assets under an Unviersal Profile.

Reviewer: ```suggestion
description: Learn how to register Issued Assets under an Universal Profile.
```

---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import Details from '../../../src/components/Details'

# Set LSP8 NFT Metadata

> In this guide you will learn how to add a Digital Asset (either [LSP7 Digital Asset](../../standards/tokens/LSP7-Digital-Asset.md) or [LSP8 Identifiable Digital Asset](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md)) to the LSP12 Issued Assets of a Universal Profile.

Reviewer: ```suggestion
In this guide you will learn how to add a Digital Asset (either [LSP7 Digital Asset](../../standards/tokens/LSP7-Digital-Asset.md) or [LSP8 Identifiable Digital Asset](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md)) inside the list of LSP12 Issued Assets of a Universal Profile.
```

## Setup

You will need the address of an existing LSP7 or LSP8 Digital Asset in order to follow this tutorial. E.g:

- LSP7 Digital Asset [`0xf7056bdE90f494F55967858F1e9E4AFB1026C5C8`](https://explorer.execution.testnet.lukso.network/address/0xf7056bdE90f494F55967858F1e9E4AFB1026C5C8?tab=read_contract).
- LSP8 Digital Asset [`0xf651b88925C0B6C81Ad6f658a2F104226d837F60`](https://explorer.execution.testnet.lukso.network/address/0xf651b88925C0B6C81Ad6f658a2F104226d837F60?tab=read_contract).

The following code snippets require the installation of the following libraries:

- [`ethers.js`](https://github.com/ethers-io/ethers.js/) or [`web3.js`](https://www.npmjs.com/package/web3)
- [`@lukso/lsp-smart-contracts`](https://github.com/lukso-network/lsp-smart-contracts/)
- [`@erc725/erc725.js`](https://github.com/ERC725Alliance/erc725.js/)

<Tabs groupId="web3-lib">
  <TabItem value="web3js" label="web3.js">

```shell
npm install web3 @lukso/lsp-smart-contracts @erc725/erc725.js
```

  </TabItem>
  <TabItem value="ethersjs" label="ethers.js">

```shell
npm install ethers @lukso/lsp-smart-contracts @erc725/erc725.js
```

  </TabItem>
</Tabs>

## Imports and constants

> Import `web3.js`/`ethers`, the [`LSP8IdentifiableDigitalAsset`](../../contracts/contracts/LSP8IdentifiableDigitalAsset/LSP8IdentifiableDigitalAsset.md) ABI from [`@lukso/lsp-smart-contracts`](../../contracts/introduction.md) and create an instance of this contract with the `lsp8ContractAddress`.

Reviewer 2: You are not using LSP8IdentifiableDigitalAsset below. Only the Universal Profile artifact and its address. This text has to change, likely on both files.

Reviewer: Correct agree with @fhildeb unless I am missing something

const issuedAssets = [
  {
    interfaceId: INTERFACE_IDS.LSP7DigitalAsset,
    address: '0xf7056bdE90f494F55967858F1e9E4AFB1026C5C8',
  },
  {
    interfaceId: INTERFACE_IDS.LSP8IdentifiableDigitalAsset,
    address: '0xf651b88925C0B6C81Ad6f658a2F104226d837F60',
  },
  //   {
  //     interfaceId: LSP7 or LSP8 interface ID (or other),
  //     address: '0xasset-address',
  //   },
];

const web3 = new Web3(window.lukso);
await web3.eth.requestAccounts();
const accounts = await web3.eth.getAccounts();
```

  </TabItem>
  <TabItem value="ethersjs" label="ethers.js">

```javascript
import { ethers } from 'ethers';

import { ERC725 } from '@erc725/erc725.js';
import LSP12Schema from '@erc725/erc725.js/schemas/LSP12IssuedAssets.json';

import UniversalProfileArtifact from '@lukso/lsp-smart-contracts/artifacts/UniversalProfile.json';
import { INTERFACE_IDS } from '@lukso/lsp-smart-contracts';

// We will register the issued assets by setting the following LSP12 data keys
// - LSP12IssuedAssets[]
// - LSP12IssuedAssetsMap:<asset-address>

// add the type of asset (LSP7 or LSP8) and their address in the object list below

> const issuedAssets = [

Reviewer 2: In the guide, you set the issued assets for the first time, but we should explain/include how to add more assets- in case the issued assets and issued assets map were not empty. Meaning people would have to get the original content, decode it, and add the new asset to the list, right?

This should be done in both examples.

Reviewer: > In the guide, you set the issued assets for the first time, but we should explain/include how to add more assets- in case the issued assets and issued assets map were not empty. Meaning people would have to get the original content, decode it, and add the new asset to the list, right?
> 
> This should be done in both examples.

erc725.js does not enable this now. It sets only from scratch. We can add a callout warning at the top, and update the guides once we have this feature

  {
    interfaceId: INTERFACE_IDS.LSP7DigitalAsset,
    address: '0xf7056bdE90f494F55967858F1e9E4AFB1026C5C8',
  },
  {
    interfaceId: INTERFACE_IDS.LSP8IdentifiableDigitalAsset,
    address: '0xf651b88925C0B6C81Ad6f658a2F104226d837F60',
  },
  //   {
  //     interfaceId: LSP7 or LSP8 interface ID (or other),
  //     address: '0xasset-address',
  //   },
];

const provider = new ethers.BrowserProvider(window.lukso);
await provider.send('eth_requestAccounts', []);
const myWallet = await provider.getSigner();
```

  </TabItem>
</Tabs>

## Encode the data keys

> Use `erc725.js` nd LSP12 JSON schema to encode the data keys related to `LSP12IssuedAssets[]`.

Reviewer 2: ```suggestion
Next, you can use the `erc725.js` library and [LSP12](../../standards/universal-profile/lsp12-issued-assets/) JSON schema to encode the data keys related to `LSP12IssuedAssets[]`.
```

Reviewer: ```suggestion
Use `erc725.js` and the LSP12 JSON schema to encode the data keys related to `LSP12IssuedAssets[]`.
```

```javascript
const erc725 = new ERC725(
  LSP12Schema,
  UNIVERSAL_PROFILE_ADDRESS,
  RPC_ENDPOINT,
  {
    ipfsGateway: 'https://api.universalprofile.cloud/ipfs',
  },
);

const allAssetAddresses = issuedAssets.map((asset) => asset.address);

const issuedAssetsMap = issuedAssets.map((asset, index) => {
  return {
    keyName: 'LSP12IssuedAssetsMap:<address>',
    dynamicKeyParts: asset.address,

>     value: [asset.interfaceId, ERC725.encodeValueType('uint128', index)],

Reviewer: Might be good for the user to add a little comment to explain what `encodeValueType('uint128', index)` does

  };
});

const { keys: lsp12DataKeys, values: lsp12Values } = erc725.encodeData([
  { keyName: 'LSP12IssuedAssets[]', value: allAssetAddresses },
  ...issuedAssetsMap,
]);
```

## Instantiate the Universal Profile contract

Create an instance of the Universal Profile contract as shown below:

<Tabs groupId="web3-lib">
  <TabItem value="web3js" label="web3.js">

```javascript
const myUPContract = new web3.eth.Contract(
  UniversalProfileArtifact.abi,
  UNIVERSAL_PROFILE_ADDRESS,
);
```

  </TabItem>
  <TabItem value="ethersjs" label="ethers.js">

```javascript
const myUPContract = new ethers.Contract(
  UNIVERSAL_PROFILE_ADDRESS,
  UniversalProfileArtifact.abi,
  myWallet,
);
```

  </TabItem>
</Tabs>

## Set data batch

> At this point we will use the data keys generated and the [step 2](#encode-the-data-keys) and Universal Profile contract instance created at [step 3](#instantiate-the-universal-profile-contract). We can use the `setDataBatch(...)` function of the Unviersal Profile to update multiple data keys.

Reviewer: ```suggestion
At this point we will use the data keys generated in [step 2](#encode-the-data-keys) and the Universal Profile contract instance created in [step 3](#instantiate-the-universal-profile-contract). We can use the `setDataBatch(...)` function of the Universal Profile to update multiple data keys.
```

Reviewer: I would even go to the point of removing the first sentence:

```suggestion
Next, use the `setDataBatch(...)` function of the Universal Profile to update multiple data keys.
```
